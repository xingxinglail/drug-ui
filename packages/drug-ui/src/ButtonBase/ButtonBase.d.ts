import * as React from 'react';
export interface ButtonBaseProps {
    className?: string;
    href?: string;
    disabled?: boolean;
    disableRipple?: boolean;
    centerRipple?: boolean;
    children?: React.ReactNode;
    component?: React.ElementType;
}
declare const ButtonBase: React.FC<ButtonBaseProps>;
export default ButtonBase;
