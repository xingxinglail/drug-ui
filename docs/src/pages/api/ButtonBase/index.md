---
filename: /packages/drug-ui/src/ButtonBase/ButtonBase.js
---

# ButtonBase API

<p class="description">ButtonBase React 组件的 API 文档。</p>

## Import

```js
import ButtonBase from '@drug-ui/core/ButtonBase';
// or
import { ButtonBase } from '@drug-ui/core';
```

```ButtonBase``` 包含尽可能少的样式。它旨在成为创建按钮的简单构建块。它包含样式重置和一些焦点/波纹逻辑的负载。

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name">centerRipple</span> | <span class="prop-type">bool</span> | false | 如果为```true```，则波纹将居中。 |
| <span class="prop-name">children</span> | <span class="prop-type">node</span> |  | 组件内容。 |
| <span class="prop-name">component</span> | <span class="prop-type">elementType</span> | <span class="prop-default">'button'</span> | 根节点组件。使用DOM元素的字符串或组件。 |
| <span class="prop-name">disabled</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 禁用按钮。 |
| <span class="prop-name">disableRipple</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 如果为```true```，则纹波效果将被禁用。 |
`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiButtonBase`.

## Demos

- [Buttons](/components/buttons/)




