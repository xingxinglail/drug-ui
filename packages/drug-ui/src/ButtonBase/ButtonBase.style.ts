import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const root: Style = {
    color: 'inherit',
    border: 0,
    cursor: 'pointer',
    margin: 0,
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    outline: 0,
    padding: 0,
    position: 'relative',
    userSelect: 'none',
    borderRadius: 0,
    verticalAlign: 'middle',
    textDecoration: 'none',
    backgroundColor: 'transparent',
    '-moz-appearance': 'none',
    '-webkit-appearance': 'none',
    '-webkit-tap-highlight-color': 'transparent',
    '&$disabled': {
        cursor: 'default',
        pointerEvents: 'none'
    }
};

const disabled: Style = {};

export const styles = (_: Theme): Styles => ({
    root: root,
    disabled
});
