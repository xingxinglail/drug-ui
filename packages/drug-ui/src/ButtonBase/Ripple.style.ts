import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const root: Style = {
    '@global': {
        '@keyframes drug-ripple-enter': {
            '0%': {
                opacity: 0.1,
                transform: 'scale(0)'
            },

            '100%': {
                opacity: 0.3,
                transform: 'scale(1)'
            }
        },
        '@keyframes drug-ripple-leave': {
            '0%': {
                opacity: 1
            },

            '100%': {
                opacity: 0
            }
        }
    },
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    zIndex: 0,
    overflow: 'hidden',
    borderRadius: 'inherit'
};

const visible: Style = {
    position: 'absolute',
    animation: '$drug-ripple-enter 550ms cubic-bezier(0.4, 0, 0.2, 1) forwards'
};

const child: Style = {
    display: 'block',
    borderRadius: '50%',
    width: '100%',
    height: '100%',
    backgroundColor: 'currentColor'
};

const childLeaving: Style = {
    animation: '$drug-ripple-leave 550ms cubic-bezier(0.4, 0, 0.2, 1) forwards'
};

export const styles = (_: Theme): Styles => ({
    root,
    visible,
    child,
    childLeaving
});
