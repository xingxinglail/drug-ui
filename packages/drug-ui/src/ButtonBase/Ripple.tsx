import * as React from 'react';
import * as PropTypes from 'prop-types';
import { createUseStyles } from '@drug-ui/styles';
import { styles } from './Ripple.style';
import { Theme } from '../styles';

interface RippleProps {
    center?: boolean;
}

type RippleClassProps = 'root' | 'visible' | 'child' | 'childLeaving';

const name = 'Ripple';
const useStyles = createUseStyles<Theme, RippleClassProps>(styles, { name });

const getRippleSize = (width: number, height: number, positionX: number, positionY: number) => {
    // 得到点击位置距离最远的点，计算出这两点之间的距离
    const square = (x: number) => x * x;
    const coordinates = [[0, 0], [width, 0], [0, height], [width, height]].map(coordinate => {
        return Math.sqrt(square(coordinate[0] - positionX) + square(coordinate[1] - positionY));
    });
    // 最长边即为半径
    const radius = Math.max.apply({}, coordinates);
    const size = radius * 2;
    const left = positionX - size / 2;
    const top = positionY - size / 2;
    return { size, left, top };
};

const Ripple: React.FC<RippleProps> = (props) => {
    const ref = React.useRef<HTMLElement>(null);
    let childDom: HTMLSpanElement;
    const classes = useStyles();
    const handleMouseDown = (e: MouseEvent) => {
        const root = ref.current;
        if (root) {
            const { left: _left, top: _top, width, height } = root.getBoundingClientRect();
            const spanDom = document.createElement('span');
            spanDom.className = classes.visible;
            let { size, left, top } = getRippleSize(width, height, e.pageX - _left, e.pageY - _top);
            if (props.center) {
                size = Math.max(width, height);
                left = (width - size) / 2;
                top = (height - size) / 2;
            }
            spanDom.style.width = `${ size }px`;
            spanDom.style.height = `${ size }px`;
            spanDom.style.top = `${ top }px`;
            spanDom.style.left = `${ left }px`;
            childDom = document.createElement('span');
            childDom.className = classes.child;
            spanDom.appendChild(childDom);
            root.appendChild(spanDom);
            const remove = () => {
                childDom.removeEventListener('animationend', remove);
                spanDom.remove();
            };
            childDom.addEventListener('animationend', remove);
        }
    };

    const handleMouseUp = () => {
        if (childDom) childDom.classList.add(classes.childLeaving);
    };

    return React.createElement('span', {
        className: classes.root,
        ref,
        onMouseDown: handleMouseDown,
        onMouseUp: handleMouseUp,
        onMouseLeave: handleMouseUp,
        onDragLeave: handleMouseUp
    });
};

Ripple.defaultProps = {
    center: false
};

Ripple.propTypes = {
    center: PropTypes.bool
};

export default Ripple;
