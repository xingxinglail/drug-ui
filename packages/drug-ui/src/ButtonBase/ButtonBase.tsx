import * as React from 'react';
import * as PropTypes from 'prop-types';
import classnames from 'classnames';
import { styles } from './ButtonBase.style';
import { createUseStyles } from '../styles';
import Ripple from './Ripple';

export interface ButtonBaseProps {
    className?: string;
    href?: string;
    disabled?: boolean;
    disableRipple?: boolean;
    centerRipple?: boolean;
    children?: React.ReactNode,
    component?: React.ElementType;
}

type ButtonBaseClassProps = 'root' | 'disabled';
const name = 'ButtonBase';
const useStyles = createUseStyles<ButtonBaseClassProps>(styles, name);

const ButtonBase: React.FC<ButtonBaseProps> = React.forwardRef<HTMLButtonElement, ButtonBaseProps>((props, ref) => {
    const { className, component = 'button', disabled, centerRipple, disableRipple, children, ...rest } = props;
    const classes = useStyles();
    const classNames = classnames(
        classes.root,
        {
            [classes.disabled]: disabled
        },
        className
    );

    let ComponentProp = component;

    if (ComponentProp === 'button' && rest.href) {
        ComponentProp = 'a';
    }

    return (
        <ComponentProp
            className={ classNames }
            ref={ ref }
            disabled={ disabled }
            { ...rest }>
            { children }
            { !disableRipple ? <Ripple center={ centerRipple } /> : null }
        </ComponentProp>
    );
});

ButtonBase.displayName = name;
ButtonBase.defaultProps = {
    className: undefined,
    disabled: false,
    disableRipple: false,
    centerRipple: false,
    component: 'button'
};

ButtonBase.propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    disableRipple: PropTypes.bool,
    centerRipple: PropTypes.bool
};

export default ButtonBase;
