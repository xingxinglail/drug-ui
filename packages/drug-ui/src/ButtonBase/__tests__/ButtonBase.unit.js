import * as renderer from 'react-test-renderer';
import React from 'react';
import { mount } from 'enzyme';
import ButtonBase from '../ButtonBase';

describe('<ButtonBase />', () => {

    it('render successful', () => {
        const json = renderer.create(<ButtonBase />).toJSON();
        expect(json).toMatchSnapshot();
    });

    it('set props className', () => {
        const wrapper = mount((
            <ButtonBase className="button button2" />
        ));
        expect(wrapper.hasClass('button button2')).toBeTruthy();
    });

    it('set props href', () => {
        const wrapper = mount((
            <ButtonBase href="https://google.com" />
        ));
        const dom = wrapper.getDOMNode();
        expect(dom.tagName).toBe('A');
        expect(dom.getAttribute('href')).toBe('https://google.com');
    });

    it('span set props href', () => {
        const wrapper = mount((
            <ButtonBase component="span" href="https://google.com" />
        ));
        const dom = wrapper.getDOMNode();
        expect(dom.tagName).toBe('SPAN');
        expect(dom.getAttribute('href')).toBe('https://google.com');
    });

    it('set props component href to aLink', () => {
        const wrapper = mount((
            <ButtonBase href="https://google.com" />
        ));
        const dom = wrapper.getDOMNode();
        expect(dom.tagName).toBe('A');
        expect(dom.getAttribute('href')).toBe('https://google.com');
    });

    it('set props component', () => {
        const Comp = ((props) => {
            const { children, ...rest } = props;
            return <a href="https://baidu.com" { ...rest }>{ children }</a>;
        });
        const wrapper = mount((
            <ButtonBase component={ Comp } />
        ));
        const dom = wrapper.getDOMNode();
        expect(dom.tagName).toBe('A');
        expect(dom.getAttribute('href')).toBe('https://baidu.com');
    });

    it('set props disabled', () => {
        const fn = jest.fn();
        const wrapper = mount((
            <ButtonBase disabled onClick={ fn } />
        ));
        const dom = wrapper.getDOMNode();
        expect(dom.getAttribute('disabled')).not.toBeNull();
        wrapper.simulate('click');
        expect(fn).not.toBeCalled();
        console.log(22);
    });

    it('set props ripple', () => {
        const wrapper = mount((
            <ButtonBase />
        ));
        const dom = wrapper.getDOMNode();
        expect(wrapper.find('span').exists()).toBeTruthy();
        wrapper.setProps({ disableRipple: true });
        expect(wrapper.find('span').exists()).toBeFalsy();
    });
});
