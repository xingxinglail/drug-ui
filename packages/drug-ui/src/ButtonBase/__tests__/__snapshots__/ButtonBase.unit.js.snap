// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonBase /> render successful 1`] = `
<button
  className="ButtonBase-root-0-2-1"
  disabled={false}
>
  <span
    className="Ripple-root-0-2-3"
    onDragLeave={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  />
</button>
`;
